sfaira/estimators/keras.py

```diff
@@ -556,75 +559,43 @@ def _get_dataset(
         if idx is None:
             idx = np.arange(0, self.data.n_obs)
 
-        if mode == 'train' or mode == 'train_val':
+        if mode in ['train', 'train_val', 'eval', 'predict']:
             # Prepare data reading according to whether anndata is backed or not:
-            if self.data.isbacked:
-                n_features = self.data.X.shape[1]
-                output_types, output_shapes = self._get_output_dim(n_features, model_type)
-
-                if model_type == "vae":
-                    def generator():
-                        sparse = isinstance(self.data.X[0, :], scipy.sparse.spmatrix)
-                        for i in idx:
-                            # (_,_), (_,sf) is dummy for kl loss
-                            x = self.data.X[i, :].toarray().flatten() if sparse else self.data.X[i, :].flatten()
-                            sf = self._prepare_sf(x=x)[0]
-                            yield (x, sf), (x, sf)
-                else:
-                    def generator():
-                        sparse = isinstance(self.data.X[0, :], scipy.sparse.spmatrix)
-                        for i in idx:
-                            x = self.data.X[i, :].toarray().flatten() if sparse else self.data.X[i, :].flatten()
-                            sf = self._prepare_sf(x=x)[0]
-                            yield (x, sf), x
-            else:
-                x = self._prepare_data_matrix(idx=idx)
-                sf = self._prepare_sf(x=x)
-                n_features = x.shape[1]
-                output_types, output_shapes = self._get_output_dim(n_features, model_type)
-
-                if model_type == "vae":
-                    def generator():
-                        for i in range(x.shape[0]):
-                            # (_,_), (_,sf) is dummy for kl loss
-                            yield (x[i, :].toarray().flatten(), sf[i]), (x[i, :].toarray().flatten(), sf[i])
-                else:
-                    def generator():
-                        for i in range(x.shape[0]):
-                            yield (x[i, :].toarray().flatten(), sf[i]), x[i, :].toarray().flatten()
+            x = self.data.X if self.data.isbacked else self._prepare_data_matrix(idx=idx)
+
+            def generator():
+                is_sparse = isinstance(x[0, :], scipy.sparse.spmatrix)
```

Comment:
@davidsebfischer I think the dtype here could also be anndata sparse object. Let me check if that's the case.

Reply:
I checked, this works for all cases.